### PR TITLE
Add UnmanagedCallersOnly attribute to SafeDeleteSslContext.ReadFromConnection/WriteToConnection methods

### LINKED ISSUE #55736: Mono: AOT compiler can't compile System.Net.SafeDeleteSslContext:ReadFromConnection (Mac Catalyst)
### Description

Calling the method `System.Net.SafeDeleteSslContext:ReadFromConnection` in AOT compiled code leads to:

> ExecutionEngineException: Attempting to JIT compile method '(wrapper native-to-managed) in System.Net.SafeDeleteSslContext:ReadFromConnection (void*,byte*,void**)' while running in aot-only mode.

This is the complete stack trace from our test:

```
[FAIL] TestNSUrlSessionHandlerCookies :   Exception
  Expected: null
  But was:  <System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
 ---> System.ExecutionEngineException: Attempting to JIT compile method '(wrapper native-to-managed) in System.Net.SafeDeleteSslContext:ReadFromConnection (void*,byte*,void**)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
   at System.Net.SafeDeleteSslContext..ctor(SafeFreeSslCredentials credential, SslAuthenticationOptions sslAuthenticationOptions) in System.Net.Security.dll:token 0x6000191+0xfc
   at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteSslContext& context, ReadOnlySpan`1 inputBuffer, Byte[]& outputBuffer, SslAuthenticationOptions sslAuthenticationOptions) in System.Net.Security.dll:token 0x6000407+0x10
 --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.<ForceAuthenticationAsync>d__172`1[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext() in System.Net.Security.dll:token 0x600030e+0x1f6
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x600051e+0x95
 --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x600051e+0x138
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x60006ff+0x410
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x6000701+0x8f
   at System.Net.Http.HttpConnectionPool.SendUsingHttp11Async(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x60006f1+0x15d
   at System.Net.Http.HttpConnectionPool.DetermineVersionAndSendAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x60006f2+0x224
   at System.Net.Http.HttpConnectionPool.SendAndProcessAltSvcAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x60006f3+0x8e
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in System.Net.Http.dll:token 0x60006f4+0x99
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken) in System.Net.Http.dll:token 0x600022f+0x1b7
   at MonoTests.System.Net.Http.MessageHandlerTest.<>c__DisplayClass4_0.<<TestNSUrlSessionHandlerCookies>b__0>d.MoveNext() in xamarin-macios/tests/monotouch-test/System.Net.Http/MessageHandlers.cs:line 106>
   at MonoTests.System.Net.Http.MessageHandlerTest.TestNSUrlSessionHandlerCookies() in xamarin-macios/tests/monotouch-test/System.Net.Http/MessageHandlers.cs:line 123
```

This is the test code: https://github.com/xamarin/xamarin-macios/blob/f5896c14141de86a377044bb0a66bbf8f982a1bb/tests/monotouch-test/System.Net.Http/MessageHandlers.cs#L106

Binlog for the test app: [build-MacCatalyst-20210715_143948.binlog.zip](https://github.com/dotnet/runtime/files/6823188/build-MacCatalyst-20210715_143948.binlog.zip)

This is the part of the log that corresponds to the AOT compilation: https://gist.github.com/rolfbjarne/fc9f1316a92945fdde0f38ff1ede334f

### PR DESCRIPTION
**Pull request by MaximLipnin**
Fixes https://github.com/dotnet/runtime/issues/55736

### CODE REVIEW

**Comment by mdh1418** [comment]
I'm not very familiar, but based on what @lambdageek mentioned https://github.com/dotnet/runtime/issues/55736#issuecomment-880690393, [UnmanagedCallersOnlyAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute?view=net-5.0) the changes seem like the right path. Are `void*` and `void**` [blittable](https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types)?

Perhaps once `Argument 2: cannot convert from '&method group' to 'delegate* unmanaged<void*, byte*, void**>'` and `An object reference is required for the non-static field, method, or property 'SafeDeleteSslContext._outputBuffer'` are resolved we'll know more?

File: src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs [hunk 1]
```diff
@@ -119,10 +119,10 @@ internal enum PAL_TlsIo
         private static extern int AppleCryptoNative_SslSetAcceptClientCert(SafeSslHandle sslHandle);
 
         [DllImport(Interop.Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_SslSetIoCallbacks")]
-        internal static extern int SslSetIoCallbacks(
+        internal static extern unsafe int SslSetIoCallbacks(
             SafeSslHandle sslHandle,
-            SSLReadFunc readCallback,
-            SSLWriteFunc writeCallback);
+            delegate* unmanaged<void*, byte*, void**> readCallback,
+            delegate* unmanaged<void*, byte*, void**> writeCallback);
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:125]:
```suggestion
            delegate* unmanaged<void*, byte*, void**, int> readCallback,
            delegate* unmanaged<void*, byte*, void**, int> writeCallback);
```
This should fix the build break

**Comment by jkotas** [comment]
> An object reference is required for the non-static field, method, or property 'SafeDeleteSslContext._outputBuffer

You will need to have some way to go from the call arguments back to the object. Do these APIs have a way to pass pointer to user data around? If not, you will need to pass the extra data around using some sort of global state, e.g. hashtable that uses connection pointer as the key.

**Comment by lambdageek** [comment]
In theory that `void* connection` is the user data.  You have to call https://developer.apple.com/documentation/security/1398843-sslsetconnection to assign a connection value to an SSLContext (somewhere before https://github.com/dotnet/runtime/blob/4a4f6c0c305331db3918007c34f6fe2836f03934/src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs#L44-L47) and then the Apple API will pass it back to the read and write callbacks.

We could probably just put a GCHandle to `this` `SafeDeleteSslContext` into the connection value.  No need for a dictionary.

**Comment by steveisok** [comment]
@MaximLipnin I put up a commit that should help.  Please validate.  

Thanks for the tips everyone.

**Review by lambdageek — COMMENTED** [review]
Nice!

**Comment by steveisok** [comment]
@MaximLipnin You might want to merge w/ main b/c it'll resolve a bunch of test failures on iOS.

**Review by lambdageek — APPROVED** [review]

**Pull request merged**